### ISSUE TITLE
Support for Upcoming Tailwind v4

### ISSUE THREAD

**Issue by allenwyma** [phenixdigital/phoenix_storybook#522]
Tailwind v4 is making some pretty big changes that seem to be for the better, especially with the speed improvements made, but they are doing away with the tailwind config file.

It looks like they will not support the recommended method of using the important config setting, is there a workaround or something else we can do to move onto Tailwind v4?

**Comment by cblavier** [comment]
Are you sure `important` won't be supported anymore?
I stumbled upon this PR: https://github.com/tailwindlabs/tailwindcss/pull/14448

**Comment by cblavier** [comment]
yes, it is ambiguous, I'll keep an eye on the beta and see how things go

**Comment by allenwyma** [comment]
I didn'tr see any option for the important, but seems the config file is support.ed

**Comment by christianRakete** [comment]
Is there any news on this topic yet?

**Comment by Coffei** [comment]
Hi, not sure if 1. is possible without 2. I wanted to upgrade our internal UI components library to Tailwind v4 but run into some issues. It seems the `app.css` included by storybook itself can conflict with the my CSS styles. E.g. consider a story with this variation.

```
      %Variation{
        id: :test,
        template: """
        <input class="border outline-none focus:ring-4" value="test"/>
        """
      }
```

If I focus the input there shoud be a ring around the input, but there is nothing. If I copy/paste this into [play.tailwindcss.com](https://play.tailwindcss.com/) it works just fine.

Upon closer inspection, the variable `--tw-ring-shadow` that is used in `ring-4` is reset to `0 0 #0000` in the storybook `app.css`, as tailwind re-sets these at the beginning, and this doesn't seem to be affected by the sandboxing class.

```
*, ::after, ::before {
	...
	--tw-ring-shadow: 0 0 #0000;
	...
}
```

Is there anything we can do about this until the phoenix_storybook is completely rewritten to Tailwind 4?

**Comment by cblavier** [comment]
I'm about to release new versions of the storybook (and the demo app) to show how we can use the storybook (still in Tailwind v3) from a Tailwind 4.x app.
This is a bit tough, I faced a lot of CSS precedence issues, and I hope it will work for most of you 🤞

**Comment by xxdavid** [comment]
Hi Christian, thank you for all your hard work!

I've tried the updated version of the demo app and it seems to work in most cases, but there is still the issue that @Coffei described, i.e., the CSS variables from Tailwind 3 and Tailwind 4 can conflict.

Take a look at this small reproducer: https://github.com/xxdavid/phoenix_storybook_demo/tree/css_var_conflicts_repro. The element `<input class="border outline-none ring-4" value="test"/>` should look like [this](https://play.tailwindcss.com/KcDnFvg97R):

![Image](https://github.com/user-attachments/assets/ac5b7e4a-b82c-4cbb-9e13-0130d9c9eac2)

but in Storybook it looks like this:

![Image](https://github.com/user-attachments/assets/a0da8cd7-e92c-41fa-96f6-cc117dbcaef9)

The `ring-4` class does not fully apply. This is because the class `ring-4` of Tailwind 4 uses a CSS variable `--tw-ring-shadow` that gets overwritten by Tailwind 3 as Tailwind 3 does not use native layers and thus has precedence over the `utilities` layer of Tailwind 4.

![Image](https://github.com/user-attachments/assets/3cb98a7b-ea43-4c15-8d64-1cf692e863e5)

![Image](https://github.com/user-attachments/assets/e78941e3-0d0a-4758-8f33-87edd85d8f2d)

Similarly, any other Tailwind 4 class that uses CSS variables that are reset in Tailwind 3 can break in Storybook. For example, neither of these work: `rotate-x-50 rotate-y-45`, `border-spacing-2`, `shadow-lg shadow-cyan-500/50`. Hopefully, not much more as the list of conflicting variables is fixed. But it's still quite problematic – one cannot trust what Storybook shows, as it can look differently in production environment.

Do you have any idea what we can do about this?

**Comment by cblavier** [comment]
Thanks for the explanation

My best idea would be to finish the work and upgrade the storybook itself to Tailwind 4.x (which would probably be a breaking change for most tailwind 3 users)

WDYT?

**Comment by cblavier** [comment]
I learned from José and Chris that maintaining compatibility with legacy dependencies is a headache best avoided. I believe moving forward is the best approach, and Tailwind 3.x users can continue using earlier versions of phoenix_storybook.

**Comment by cblavier** [comment]
Seems to be working now 🤞
Does anyone want to test branch `dependabot/npm_and_yarn/assets/tailwindcss-4.1.11`?
PS: documentation hasn't been yet updated

**Comment by xxdavid** [comment]
Hi, I gave it a quick try and everything seems to be working well! Thank you so much for updating Storybook to Tailwind 4! 🙂
